tests/substrate.rs

```diff
@@ -1,1145 +1,866 @@
+        hasher.update(&read_buf(mem, input_ptr, input_len));
+        hasher.finalize(&mut mem[output_ptr as usize..(output_ptr + 32) as usize]);
+        Ok(())
+    }
 
-                std::mem::swap(&mut self.vm, &mut vm);
+    #[seal(0)]
+    fn seal_hash_sha2_256(input_ptr: u32, input_len: u32, output_ptr: u32) -> Result<(), Trap> {
+        let mut hasher = Sha256::new();
+        hasher.update(read_buf(mem, input_ptr, input_len));
+        write_buf(mem, output_ptr, &hasher.finalize());
+        Ok(())
+    }
 
-                let module = self.create_module(&program.code);
+    #[seal(0)]
+    fn seal_hash_blake2_128(input_ptr: u32, input_len: u32, output_ptr: u32) -> Result<(), Trap> {
+        let data = read_buf(mem, input_ptr, input_len);
+        write_buf(mem, output_ptr, blake2b(16, &[], &data).as_bytes());
+        Ok(())
+    }
 
-                self.vm.input = input;
+    #[seal(0)]
+    fn seal_hash_blake2_256(input_ptr: u32, input_len: u32, output_ptr: u32) -> Result<(), Trap> {
+        let data = read_buf(mem, input_ptr, input_len);
+        write_buf(mem, output_ptr, blake2b(32, &[], &data).as_bytes());
+        Ok(())
+    }
 
-                let ret = match module.invoke_export("deploy", &[], self) {
-                    Err(wasmi::Error::Trap(trap)) => match trap.kind() {
-                        TrapKind::Host(host_error) => {
-                            if let Some(ret) = host_error.downcast_ref::<HostCodeReturn>() {
-                                Some(RuntimeValue::I32(ret.0))
-                            } else {
-                                return Err(trap);
-                            }
-                        }
-                        _ => {
-                            return Err(trap);
-                        }
-                    },
-                    Ok(v) => v,
-                    Err(e) => panic!("fail to invoke deploy: {e}"),
-                };
-
-                let output = self.vm.output.clone();
-
-                std::mem::swap(&mut self.vm, &mut vm);
-
-                set_seal_value!(
-                    "seal_instantiate output",
-                    output_ptr,
-                    output_len_ptr,
-                    &output
-                );
-
-                if let Some(RuntimeValue::I32(0)) = ret {
-                    self.accounts.get_mut(&vm.account).unwrap().1 += vm.value;
-                    set_seal_value!(
-                        "seal_instantiate account",
-                        account_ptr,
-                        account_len_ptr,
-                        &account
-                    );
-                }
+    #[seal(1)]
+    fn seal_call(
+        _flags: u32,
+        callee_ptr: u32,
+        _gas: u64,
+        value_ptr: u32,
+        input_ptr: u32,
+        input_len: u32,
+        output_ptr: u32,
+        output_len_ptr: u32,
+    ) -> Result<u32, Trap> {
+        let input = read_buf(mem, input_ptr, input_len);
```

Comment:
The old code ensured that `_flags` was set to zero

Reply:
Thanks, added the check again